DM-35455: Create a composite action for building and publishing to PyPI

jonathansick opened the pull request:
This PR sets up the repository as a GitHub [Composite Action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) that can be called as a regular action step either by other repo's GitHub Actions workflows or through reusable workflows.

This action specifically packages our common practices of building a Python package's distribution and publishing it to GitHub Actions, as described in the README docs.

To build the distribution, I'm now using the [build](https://pypa-build.readthedocs.io/en/stable/index.html) tool, which is the PEP 587 modernization of `python setup.py sdist bdist wheel`. This is actually new to me. To do the PyPI upload, I'm using [pypa/gh-action-pypi-publish](https://github.com/pypa/gh-action-pypi-publish), which we've used previously.

I chose not to include `actions/checkout` in this composite action so that end-users could a) specify options for checking out the code, and b) easily run action steps before building and publishing, such as building/copying in assets or other random things.

A basic example of using this composite action:

```yaml
name: Python CI

"on":
  push:
    tags:
      - "*"
  pull_request: {}

jobs:

  pypi:

    runs-on: ubuntu-latest
    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags/')

    steps:
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0 # full history for setuptools_scm

      - name: Build and publish
        uses: lsst-sqre/build-and-publish-to-pypi@v1
        with:
          pypi-token: ${{ secrets.PYPI_SQRE_ADMIN }}
```

jonathansick committed:
Update copyright in license

jonathansick committed:
Add pre-commit for linting

This pre-commit configuration primarily ensures that the YAML is
well-configured. The check-jsonschema is particularly helpful because it
validates both workflows and the composite action source against a
schema.

NOTE: turns out the github-actions schema doesn't know how to deal with
an action.yaml that has a runs.using="composite" field, instead of one
that points to a JavaScript file. Commenting out for now.

jonathansick committed:
Create composite action for build and publish

jonathansick committed:
Add a dependabot configuration

This will help keep the actions used by this action up-to-date

jonathansick committed:
Document the action

jonathansick committed:
Add lint and release GitHub Actions workflows

rra commented on README.md:
```diff
@@ -1,2 +1,82 @@
+## Inputs
+
+- `pypi-token` (string, required) an API token for PyPI.
+- `python-version` (string, optional) the Python version. Default is `3.10`.
```
Thinking ahead here, I wonder if it would make sense to always require specifying the version. My intuition is that I would normally want the build and PyPI upload to be done with the same Python version that the package is targeting (it may not build on newer versions, for instance, or may build binary wheels that should match the supported Python version), and if it's always specified in the package GitHub Actions workflow definition, I'll remember to update it when I update the Python versions for tests, docs (if relevant), etc. But if the default of the composite action is used, I'm likely to forget that it has a default version, or what that version is, and it seems more likely that we'll end up trying to publish with a different version than the package was tested with.

That's more awkward with this action than some others, such as docs, since it only runs on tags, so if it does fail due to a version mismatch one wouldn't know until the release is already tagged and then fixing it would probably require making and tagging a new release.

jonathansick replied:
Hm, this is an interesting question. My original thinking was we'd update the Python version in major releases of the action so that dependabot would effectively help us keep up to date by default.

However, I think you're right that we'll always probably need to keep an eye on our Python versions. For example, I think the best way to do a matrix of tox-based tests would be to wrap up tox in a composite action and let the package's workflow then control what versions of Python it cares to test against.

rra commented on README.md:
```diff
@@ -1,2 +1,82 @@
+
+## Developer guide
+
+This repository enables us to centrally maintain our GitHub Actions workflows through centrally-maintained [composite actions](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action).
```
Nit: The phrasing here implies that we could put multiple different composite actions in this same repository, but the name of the repository seems to imply that it only provides one workflow (and the syntax of `action.yaml` also seems to imply that). Maybe this should be singular in some places where it's currently plural ("composite actions" for example)?

rra reviewed (approved):
Thanks for doing this! Really excited to use more of these.

jonathansick committed:
Make the python-version required

This makes the Python version more transparent for end-users, and
prevents us from tying up major new releases of this action with major
releases of Python itself.
